Add gateway collaborators management

kschiffer opened the pull request:
#### Summary
Closes #777

#### Changes
- Add collaborator service to gateway service (JS SDK)
- Update api tree accordingly
- Add gateway collaborator views and routes
- Refactor `<RightsGroup />` to use an array of rights as prop and remove resulting unnecessary marshalling
- Fix related selector usage

#### Release Notes
Added gateway collaborators page to console

bafonins commented on pkg/webui/console/views/gateway-collaborator-add/gateway-collaborator-add.styl:
```diff
@@ -0,0 +1,18 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+.right-label
```
just a remainder
after https://github.com/TheThingsNetwork/lorawan-stack/pull/817 it wont be needed

bafonins commented on pkg/webui/console/views/gateway-collaborator-add/index.js:
```diff
@@ -0,0 +1,206 @@
+})
+
+@connect(function ({ rights, collaborators }, props) {
+  const gtwId = props.match.params.gtwId
```
please add `selectSelectedGatewayId` selector

bafonins commented on pkg/webui/console/views/gateway-collaborator-add/index.js:
```diff
@@ -0,0 +1,206 @@
+  return {
+    gtwId,
+    collaborators: collaborators.gateways.collaborators,
+    fetching: rights.gateways.fetching,
```
you can use existing selectors for rights

bafonins commented on pkg/webui/console/views/gateway-collaborator-add/index.js:
```diff
@@ -0,0 +1,206 @@
+  componentDidMount () {
+    const { dispatch, gtwId } = this.props
+
+    dispatch(getGatewaysRightsList(gtwId))
```
move this to `mapDispatchToProps`

bafonins commented on pkg/webui/console/views/gateway-collaborator-edit/gateway-collaborator-edit.styl:
```diff
@@ -0,0 +1,17 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+.right-label
```
just a remainder
after https://github.com/TheThingsNetwork/lorawan-stack/pull/817 it wont be needed

bafonins commented on pkg/webui/console/views/gateway-collaborator-edit/index.js:
```diff
@@ -0,0 +1,268 @@
+import style from './gateway-collaborator-edit.styl'
+
+// TODO: Move this to checkbox group later, see https://github.com/TheThingsNetwork/lorawan-stack/issues/189
+const UNIVERSAL_APPLICATION_RIGHTS = [ 'RIGHT_APPLICATION_ALL', 'RIGHT_ALL' ]
```
just a remainder
after https://github.com/TheThingsNetwork/lorawan-stack/pull/817 it wont be needed

bafonins commented on pkg/webui/console/views/gateway-collaborator-edit/index.js:
```diff
@@ -0,0 +1,268 @@
+})
+
+@connect(function ({ collaborators, rights }, props) {
+  const { gtwId, collaboratorId } = props.match.params
```
please add `selectSelectedGatewayId` selector

bafonins commented on pkg/webui/console/views/gateway-collaborators-list/index.js:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  baseDataSelector ({ collaborators }) {
+    const { gtwId } = this.props.match.params
```
please add `selectSelectedGatewayId` selector

bafonins commented on pkg/webui/console/views/gateway-collaborators/index.js:
```diff
@@ -0,0 +1,55 @@
+import GatewayCollaboratorAdd from '../gateway-collaborator-add'
+import GatewayCollaboratorEdit from '../gateway-collaborator-edit'
+
+@withBreadcrumb('apps.single.api-keys', function (props) {
```
```suggestion
@withBreadcrumb('gtws.single.api-keys', function (props) {
```

bafonins commented on pkg/webui/console/views/gateway-collaborators/index.js:
```diff
@@ -0,0 +1,55 @@
+
+@withBreadcrumb('apps.single.api-keys', function (props) {
+  const { match } = props
+  const gtwId = match.params.gtwId
```
please add `selectSelectedGatewayId` selector

bafonins commented on pkg/webui/console/views/gateway-collaborator-add/index.js:
```diff
@@ -0,0 +1,206 @@
+                autoFocus
+              />
+              <Field
+                type="select"
```
we can now use radio buttons here

bafonins commented on pkg/webui/console/views/gateway-collaborator-edit/index.js:
```diff
@@ -0,0 +1,268 @@
+@connect(function ({ collaborators, rights }, props) {
+  const { gtwId, collaboratorId } = props.match.params
+  const collaboratorsFetching = collaborators.gateways.fetching
+  const rightsFetching = rights.gateways.fetching
```
you can use existing selectors for rights

kschiffer commented:
I've refactored to use a common `<CollaboratorForm />` component.
I've also added functionality to the rights group to pass universal right values, so they can be chosen to be prioritized over the individual rights in the put request. (as in: setting the universal right instead of the respective rights individually)

kschiffer commented:
I've addressed your points, except for 3. which is to my understanding the desired behavior (preferring universal rights over their corresponding individual values, if applicable).

Additionally, I have identified the following problems:
- Proper composition of rights can be non-trivial when for example entity rights are modified for a user that possesses `RIGHT_ALL`. Since the application/gateway collaborator edit views will only modify rights of this specific entity, revoking, say application rights, would mean to compose a list of rights that has all rights, except the ones for applications. The backend does to my understanding not offer any assistance to handle this. As such, I have modified the behavior of the collaborator form to prevent modifying any rights of a collaborator with `RIGHT_ALL`. Otherwise, we risk users revoking rights that they did not intend to.
- There can be users that have fewer rights themselves than an existing API key of an entity that they are a collaborator of. In these cases, I disable any right-checkbox that they are not allowed to modify with their own rights.
- I have also refactored the `<RightsGroup />`-component to receive an array of set rights as its value, reflecting the data structure of the entity in the backend. This makes handling a lot easier and gets rid of a lot of marshaling necessary before sending the form data to the API.

kschiffer commented on pkg/webui/console/components/collaborator-form/index.js:
```diff
@@ -0,0 +1,242 @@
+      onSubmitSuccess()
+    } catch (error) {
+      setSubmitting(false)
+      this.setState({ error })
```
It's not necessary to wait for the state change.

kschiffer commented on pkg/webui/console/views/application-collaborator-add/index.js:
```diff
@@ -44,28 +36,23 @@ import {
     error: applicationRightsErrorSelector(state, props),
   }
+}, function (dispatch, ownProps) {
+  const appId = ownProps.match.params.appId
```
Since I cannot access the state in there, this has to be solved using connect's `mergeProps` function.

bafonins commented on pkg/webui/console/components/collaborator-form/index.js:
```diff
@@ -0,0 +1,241 @@
+    )
+  }
+}
+
```
redundant new line

bafonins commented on pkg/webui/console/components/collaborator-form/index.js:
```diff
@@ -0,0 +1,241 @@
+          required
+          component={RightsGroup}
+          rights={rights}
+          universalRight={universalRights.length ? universalRights[0] : undefined}
```
`universalRights` should default to an empty array and you can simply pass `universalRights[0]`

bafonins commented on pkg/webui/console/components/collaborator-form/index.js:
```diff
@@ -0,0 +1,241 @@
+      })
+      onDeleteSuccess()
+    } catch (error) {
+      await this.setState({ error })
```
Is it necessary to wait for the state change here?

bafonins commented on pkg/webui/console/components/rights-group/index.js:
```diff
@@ -18,22 +18,27 @@ import { defineMessages } from 'react-intl'
 
 import PropTypes from '../../../lib/prop-types'
+import { RIGHT_ALL } from '../../lib/rights'
+
```
redundant new line

kschiffer replied:
No, we separate `lib` from component imports in other places too.

bafonins replied:
no we dont, but whatever

kschiffer replied:
After looking at it again, let's say there is at least no consistency in how we do it, we do however indeed not separate them in most cases.

https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/webui/console/containers/webhooks-table/index.js

https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/webui/components/location-form/index.js

I guess this is a case for clarification in some yet to be written coding conventions.

bafonins commented on pkg/webui/console/components/rights-group/index.js:
```diff
@@ -118,26 +146,43 @@ class RightsGroup extends React.Component {
           key={right}
           name={right}
           label={{ id: `enum:${right}` }}
+          disabled={outOfOwnScopeRights.includes(right)}
```
What is the purpose of this? I can agree that is looks nice, but the user with less rights cannot modify another collaborator that has more rights. The IS would return a 403 error. 
<img width="644" alt="Screenshot 2019-06-26 at 13 42 44" src="https://user-images.githubusercontent.com/16374166/60177104-52f5fa80-9818-11e9-869d-ba6502679230.png">

In this case the current user has only view application/edit application/view and edit app collaborators rights, while `test-user2` has more.

kschiffer replied:
Yeah, this is actually more complex than thought. The API behavior is in my eyes not optimal (see #857). Given the *edge-casyness* of this thing. I would go for just disabling altering any rights if the given collaborator has more rights than oneself.

kschiffer replied:
I have added a `strict` prop to the `<RightsGroup />` that will prevent any rights modification of users with more rights. The whole scenario is too much of an edge case to go on with it at the moment. We should pick this up later on but we need a proper strategy for communicating this constraint to the user. The logic we have now is a good start to do this,.

bafonins commented on pkg/webui/console/views/application-collaborator-edit/index.js:
```diff
@@ -83,15 +61,20 @@ const m = defineMessages({
   }
-}, dispatch => ({
+}, (dispatch, ownProps) => ({
   async loadData (appId) {
```
```suggestion
  loadData (appId) {
```
